### PR TITLE
PT-155292768 Improve docker build speed

### PR DESCRIPTION
**Pull request by amir-hadi**
This little PR has great impact on the docker image building. I've changed rebar to allow a two-step release process:
1. compile the dependencies listed in `rebar.config`
2. compile the rest of the application

One can see that in the Dockerfile I first add all files required by _rebar3_ to compile the dependencies. After the dependencies have been compiled, the rest of the app is added to the docker image and then compiled.

The source code used to compile this custom version of _rebar3_ is in [my own fork](https://github.com/amir-hadi/rebar3/commit/511ec8a08fe872d16068a87936cef02df2a8b03a). I have also created a [PR based on this commit](https://github.com/erlang/rebar3/pull/1720). It won't be merged upstream until someone more suitable has written tests for my proposed changes of _rebar3_.

The version of rebar3 in this PR is `rebar 3.5.0+build.3979.ref511ec8a0 on Erlang/OTP 20 Erts 9.2.1`.
I have used erlang/OTP 20 to compile rebar3 locally. More details can be seen here:
```{19:07}~/Workspace/aeternity/epoch:PT-155292768-improve-docker-build-speed ✓ ➭ erl
Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]

Eshell V9.2.1  (abort with ^G)
1>
```

The version of rebar3 in the repository before this PR is `rebar 3.4.7 on Erlang/OTP 20 Erts 9.2.1`. As I didn't compile that rebar3 myself I can't tell which version of OTP was used (maybe its the one mentioned by `rebar3 --version`: `Erlang/OTP 20 Erts 9.2.1`).

### CODE REVIEW

**Comment by dincho** [comment]
@lucafavatella I'll change the commit message during rebase+merge once this is approved. The commit message I plan to use is:

- the rebar3 binary was compiled from [upstream PR commit](https://github.com/erlang/rebar3/pull/1720/commits/511ec8a08fe872d16068a87936cef02df2a8b03a)
- based on version 3.5.0
- compiled with OTP: `Erlang/OTP 20 [erts-9.2.1] [source] [64-bit] [smp:8:8] [ds:8:8:10] [async-threads:10] [hipe] [kernel-poll:false] [dtrace]`
- old rebar3 version: `rebar 3.4.7 on Erlang/OTP 20 Erts 9.2.1` binary downloaded from [rebar3 GitHub releases](https://github.com/erlang/rebar3/releases/tag/3.4.7)
- full version output: `rebar 3.5.0+build.3979.ref511ec8a0 on Erlang/OTP 20 Erts 9.2.1`

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,9 @@
 FROM aetrnty/builder as builder
 
+# Add required files to download and compile only the dependencies
+ADD rebar.config rebar.lock Makefile rebar3 VERSION /app/
```
Comment (lucafavatella) [inline Dockerfile]: I see rebar.config.script not included meaning "release" not set correctly. I expect no impact for building deps now.
Though, in order to be more robust to future changes in rebar.config.script, may I suggest including it here so, if it gets to do more and impact the deps, this Dockerfile is not affected.

Comment (dincho) [reply Dockerfile]: I agree that this script might impact the dependancies in future.

@amir-hadi could you please include it?

Comment (amir-hadi) [reply Dockerfile]: @dincho Added the requested script in the first step.

**Pull request merged**